doc/man/man1/auks.1
.TH "auks" "1" "March 2009" "Matthieu Hautreux" "auks"

.SH "NAME"
auks \- client program to manage AUKS credentials

.SH "SYNOPSIS"
\fBauks\fR [\fI-h\fR] [\fI-dv\fR] [\fI-f auks.conf\fR] [\fI-p|a|g|r|D\fR] [\fI-R once|loop\fR] [\fI-u uid\fR] [\fI-C ccache\fR]

.SH "DESCRIPTION"
.LP
The AUKS commandline client can be used to manage AUKS daemon credentials.
Depending on his privileges, a client can use it to add/get/remove
credentials or dump all the available credentials.

.SH "OPTIONS"
.LP

.TP
\fB\-h\fR
print this message

.TP
\fB\-v\fR
increase verbosity output

.TP
\fB\-d\fR
increase debug output

.TP
\fB\-f auks.conf\fR
use \fIauks.conf\fR as the configuration file.
The default is \fB/etc/auks/auks.conf\fR.

.TP
\fB\-p\fR
do a simple ping request on the first responding AUKS daemon (default).
This operation requires at least to have a valid kerberos credential
to connect to the AUKS daemon

.TP
\fB\-a\fR
add client kerberos credential to the first responding AUKS daemon. If 
client kerberos credential is not an addressless kerberos TGT, an
addressless cred is gotten and then transmitted in place of the given one.
This operation requires at least the \fBguest\fR privilege.

.TP
\fB\-g\fR
get AUKS kerberos credential of the requested uid using the first responding 
AUKS daemon and stores it into the requested credential cache. This operation 
requires at least the \fBuser\fR privilege.

.TP
\fB\-r\fR
remove AUKS kerberos credential of the requested uid using the first responding 
AUKS daemon. This operation requires at least the \fBuser\fR privilege.

.TP
\fB\-D\fR
dump all the AUKS kerberos credentials using the first responding AUKS daemon.
This operation requires at least the \fBadmin\fR privilege.

.TP
\fB\-R once|loop\fR
use AUKS commandline client as a credential renewer process. In \fBloop\fR mode,
the process exits only when an error occured or the credential can no longer 
be renewed. Used renewal mechanism is to first try to renew credential using 
the first responding AUKS daemon and second to use kerberos KDC in case of 
failure. In \fBonce\fR mode, the ticket is renewed once, only if needed, and 
the program exits. This operation requires at least the \fBuser\fR privilege.

.TP
\fB\-u uid\fR
specify the requested uid to use with get/remove requests

.TP
\fB\-C ccache\fR
specify the input/output credential file to use with add/get requests


.SH "ERRORS"
.LP
exit status is zero in case of success, not zero otherwise.

.SH "ENVIRONMENT VARIABLES"
.LP
.TP
\fBKRB5CCNAME\fR
Like any kerberos client, kerberos credentials used during AUKS
communications are stored into the default cache or the cache 
pointed	to by this environment variable.

.SH "COPYING"
.LP
Copyright  CEA/DAM/DIF (2009)
.br

.br
This software is a computer program whose purpose is to simplify
.br
the addition of kerberos credential support in Batch applications.
.br

.br
This software is governed by the CeCILL-C license under French law and
.br
abiding by the rules of distribution of free software.  You can  use, 
.br
modify and/ or redistribute the software under the terms of the 
.br
CeCILL-C license as circulated by CEA, CNRS and INRIA at the 
.br
following URL "http://www.cecill.info". 
.br

.br
As a counterpart to the access to the source code and  rights to copy,
.br
modify and redistribute granted by the license, users are provided only
.br
with a limited warranty  and the software's author,  the holder of the
.br
economic rights,  and the successive licensors  have only  limited
.br
liability. 
.br

.br
In this respect, the user's attention is drawn to the risks associated
.br
with loading,  using,  modifying and/or developing or reproducing the
.br
software by the user in light of its specific status of free software,
.br
that may mean  that it is complicated to manipulate,  and  that  also
.br
therefore means  that it is reserved for developers  and  experienced
.br
professionals having in-depth computer knowledge. Users are therefore
.br
encouraged to load and test the software's suitability as regards their
.br
requirements in conditions enabling the security of their systems 
.br
and/or data to be ensured and,  more generally, to use and operate 
.br
it in the same conditions as regards security. 
.br

.br
The fact that you are presently reading this means that you have had
.br
knowledge of the CeCILL-C license and that you accept its terms.
.br

.SH "SEE ALSO"
.BR auks.conf (5),
.BR auksd (8)



ChangeLog
06/16/2020 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.5.0
	* auks.spec: add libtirpc requirements starting from Rhel8/Fedora28
	* auks-api: use libtirpc when detected at configure
	(based on a patch from Regine Gaudin, CEA)
	* auks/krb5|aukspriv|spank-auks: add compat with all krb5 ccache types
	(patch from Romain Fihue, CEA)
	* auks/krb5: fix TGT filtering method to handle non file-based ccache
	(patch from Romain Fihue, CEA)
	* spank-auks: move local ctx init to slurm_spank_init_post_opt
	* spank-auks: fix #24 slurmstepd not killing auks process
	(patch from Raouf Bencheraiet)
	* auks-api: enlarge AUKS_CRED_DATA_MAX_LENGTH to 4096 to ensure that
	TGT larger than 2ko are properly handled.
	/!\ breaks compatibility with previous versions of auks, all slurm
	components must be upgraded at the same time on all nodes.
	(patch from Robbert Eggermont, TU Delft)
	* auks: make options parsing work with unsigned char
	(patch from Raouf Bencheraiet)

04/06/2016 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auksd/systemd: set max open fds limit to 32k

02/02/2016 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* aukspriv/systemd: allow shell commands in environment file
	* Fix systemd dependencies (patch from Robbert Eggermont, TU Delft)

11/24/2015 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* slurm/spank: add 'enforced' conf options to request for
	  a valid credential at submission time or return an error
	  to the spank stack. (see man auks.so)

11/18/2015  Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.4.4-3
	* auksd/auksdrenewer: add foreground mode using -F flag
	* systemd: now use 'simple' systemd type using the foreground
	  mode and collectl logs on stderr (managed using journalctl)
	* specfile: correct a regression preventing from putting
	  initscripts in the right directory

10/19/2015  Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.4.3-3
	* integrate Systemd init scripts for auks components

03/27/2015  Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.4.3
	* configure: fix for automake 1.12 (CentOS-7.x)
	* initscripts: remove unused pidfiles in headers (CentOS-7.x)

01/26/2015 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.4.2
	* init scripts : change order to be launched after NTP on RHEL-6.x
	* slurm/spank: ensure that the renewer launched using spank is working
	  in / and waited for at the end of the step.
	* cleaner: correct a bug resulting in non renewable tickets removal
	  at each cleaning stage in auksd.

	* auks-0.4.1(-x)
	* aukspriv: add kinit timeout logic
	* auks: automatically inhibate auks support when no ccache available
	* auks SLURM Spank plugin: add hostcredcache= conf pragma to specify
	  a particular credential cache for the slurmd actions (based on a
	  patch from Nacho Barrientos, CERN)
	* code cleaning for cppcheck usage
	* auks SLURM Spank plugin: add sync=mode conf pragma (cf man auks.so)
	* libauksapi: correct a bug in uid management greater than 999999999

10/07/2010 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.4.0-1
	* correct client/server connection timeout
	* improve init service internals to prevent possible daemon hangup
	* add minimum uid configuration param to auks slurm spank plugin
	* remove memory leak workaround required by MIT kerberos version
	  prior to 1.8 (use -DLIBKRB5_MEMORY_LEAK_WORKAROUND to activate
	  this workaround)

01/28/2010 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.3.8-1 official beta release
	* slurm plugin: add the possibility to use credential
	in successive spank plugins not just in the launched
	tasks

12/09/2009 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.3.7-1 official release

12/07/2009 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.3.6-4
	* slurm plugin: correct a bug when --auks is used in slurm
	  without argument

12/02/2009 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.3.6-3
	* slurm plugin: enhance command line and conf file options
	  (enable/disable)
	* correct a bug in aukspriv that terminate the service if a
	  temporary error prevent from talking with the KDC
	* correct a bug preventing from getting a valid addressless ticket
	  from an addressfull one

10/28/2009 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.3.6-2
	* correct a bug preventing from getting a valid addressless
	ticket from an addressfull one (get a forwarded addressless ticket
	instead of getting a renewed ticket with NULL addr)

06/07/2009 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.3.6-1
	* add external renewer support to slurm spank plugin

03/27/2009 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.3.4
	* add aukspriv scripted daemon and associated service
	* add auks_api_renew_cred functionality

03/26/2009 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.3.3
	* auksdrenew daemon
	* configuration file structure enhancement

03/17/2009 Matthieu Hautreux <matthieu.hautreux@cea.fr>

	* auks-0.3.2
	* auksdrenew support creation


src/auks/auks.c
/***************************************************************************\
 * auks.c - user/admin interface to manage AUKS credentials 
 ***************************************************************************
 * Copyright  CEA/DAM/DIF (2009)
 *
 * Written by Matthieu Hautreux <matthieu.hautreux@cea.fr>
 * 
 * This software is a computer program whose purpose is to simplify
 * the addition of kerberos credential support in Batch applications.
 *
 * This software is governed by the CeCILL-C license under French law and
 * abiding by the rules of distribution of free software.  You can  use, 
 * modify and/ or redistribute the software under the terms of the CeCILL-C
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info". 
 * 
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability. 
 * 
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or 
 * data to be ensured and,  more generally, to use and operate it in the 
 * same conditions as regards security. 
 * 
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL-C license and that you accept its terms.
 ***************************************************************************
 * Copyright  CEA/DAM/DIF (2009)
 * 
 * Ecrit par Matthieu Hautreux <matthieu.hautreux@cea.fr>
 *
 * Ce logiciel est un programme informatique servant à faciliter l'ajout
 * du support des tickets Kerberos aux applications Batch.
 * 
 * Ce logiciel est régi par la licence CeCILL-C soumise au droit français et
 * respectant les principes de diffusion des logiciels libres. Vous pouvez
 * utiliser, modifier et/ou redistribuer ce programme sous les conditions
 * de la licence CeCILL-C telle que diffusée par le CEA, le CNRS et l'INRIA 
 * sur le site "http://www.cecill.info".
 * 
 * En contrepartie de l'accessibilité au code source et des droits de copie,
 * de modification et de redistribution accordés par cette licence, il n'est
 * offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
 * seule une responsabilité restreinte pèse sur l'auteur du programme,  le
 * titulaire des droits patrimoniaux et les concédants successifs.
 * 
 * A cet égard  l'attention de l'utilisateur est attirée sur les risques
 * associés au chargement,  à l'utilisation,  à la modification et/ou au
 * développement et à la reproduction du logiciel par l'utilisateur étant 
 * donné sa spécificité de logiciel libre, qui peut le rendre complexe à 
 * manipuler et qui le réserve donc à des développeurs et des professionnels
 * avertis possédant  des  connaissances  informatiques approfondies.  Les
 * utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
 * logiciel à leurs besoins dans des conditions permettant d'assurer la
 * sécurité de leurs systèmes et ou de leurs données et, plus généralement, 
 * à l'utiliser et l'exploiter dans les mêmes conditions de sécurité. 
 * 
 * Le fait que vous puissiez accéder à cet en-tête signifie que vous avez 
 * pris connaissance de la licence CeCILL-C, et que vous en avez accepté les
 * termes. 
\***************************************************************************/
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>

#include <string.h>

#include <time.h>

#include "auks/auks_error.h"
#include "auks/auks_api.h"
#include "xternal/xlogger.h"

#define PING_REQUEST    1
#define ADD_REQUEST     2
#define GET_REQUEST     3
#define REMOVE_REQUEST  5
#define DUMP_REQUEST    6
#define RENEW_REQUEST   7

#define RENEW_MODE_ONCE             1
#define RENEW_MODE_UNTIL_REMOVAL    2
#define RENEW_MODE_UNTIL_EXPIRATION 3


int
main(int argc,char** argv)
{

	int fstatus = -1 ;

	int action=0;
	char* ccache=NULL;
	uid_t requested_uid;

	int debug_level=0;
	int verbose_level=0;
	char* conf_file_string;

	/* options processing variables */
	char* progname;
	char* optstring="dvhf:H:pagrDC:u:R:";
	char* short_options_desc="\nUsage : %s [-h] [-dv] [-f conffile] \
[-C ccache] [-p|a|g|r|D] [-R once|loop] [-u uid] \n\n";
	char* addon_options_desc="\
\t-h\t\tshow this message\n\
\t-d\t\tincrease debug level\n\
\t-v\t\tincrease verbose level\n\
\t-f conffile\tConfiguration file\n\
\t-p\t\tping request (default)\n\
\t-a\t\tadd request\n\
\t-g\t\tget request\n\
\t-D\t\tdump request\n\
\t-r\t\tremove request\n\
\t-R mode\t\trenew credential according to specified mode\n\
\t-C ccache\tConfiguration file\n\
\t-u uid\t\tuid of requested cred owner (get request only)\n\n";

	int  option;
  
	auks_engine_t engine;

	auks_cred_t* creds;
	int creds_nb;

	int renew_mode;

	/* set default requested uid */
	requested_uid=geteuid();

	/* get current program name */
	progname=rindex(argv[0],'/');
	if(progname==NULL)
		progname=argv[0];
	else
		progname++;

	conf_file_string=NULL;

	/* process options */
	while((option = getopt(argc,argv,optstring)) != -1)
	{
		switch(option)
		{
		case 'v' :
			verbose_level++;
			break;
		case 'd' :
			debug_level++;
			break;
		case 'f' :
			conf_file_string=strdup(optarg);
			break;
		case 'C' :
			ccache=strdup(optarg);
			if(ccache==NULL){
				fprintf(stderr,"memory allocation error!");
				return 1;
			}
			break;
		case 'p' :
			action=PING_REQUEST;
			break;
		case 'a' :
			action=ADD_REQUEST;
			break;
		case 'g' :
			action=GET_REQUEST;
			break;
		case 'r' :
			action=REMOVE_REQUEST;
			break;
		case 'R' :
			action=RENEW_REQUEST;
			if(strncmp(optarg,"once",5)==0)
				renew_mode = AUKS_API_RENEW_ONCE ;
			else if(strncmp(optarg,"loop",5)==0)
				renew_mode = AUKS_API_RENEW_LOOP ;
			else {
				fprintf(stdout,short_options_desc,progname);
				fprintf(stdout,"%s\n",addon_options_desc);
				exit(1);
			}
			break;
		case 'D' :
			action=DUMP_REQUEST;
			break;
		case 'u' :
			requested_uid=(uid_t)atoi(optarg);
			break;
		case 'h' :
		default :
			fprintf(stdout,short_options_desc,progname);
			fprintf(stdout,"%s\n",addon_options_desc);
			exit(0);
			break;
		}
	}

	/* set verbosity and debug level */
	xdebug_setmaxlevel(debug_level);
	xverbose_setmaxlevel(verbose_level);

	/* load configuration */
	fstatus = auks_api_init(&engine,conf_file_string);
	if ( fstatus != AUKS_SUCCESS ) {
		fprintf(stderr,"Auks API init failed : %s\n",
			auks_strerror(fstatus));
		goto exit;
	}

	if ( verbose_level > 0 )
		auks_api_set_logfile(&engine,"/dev/stdout");
	
	switch(action){
		
	case ADD_REQUEST :
		fstatus = auks_api_add_cred(&engine,
					    ccache);
		break;
		
	case GET_REQUEST :
		fstatus = auks_api_get_cred(&engine,
					    requested_uid,
					    ccache);
		break;
		
	case DUMP_REQUEST :
		fstatus = auks_api_dump(&engine,&creds,&creds_nb);
		break;

	case REMOVE_REQUEST :
		fstatus = auks_api_remove_cred(&engine,
					       requested_uid);
		break;
		
	case RENEW_REQUEST :

		fstatus = auks_api_renew_cred(&engine,ccache,renew_mode);

		break;

	case PING_REQUEST :
	default :
		fstatus = auks_api_ping(&engine);
		break;

	}

	if ( fstatus != AUKS_SUCCESS ) {
		fprintf(stderr,"Auks API request failed : %s\n",
			auks_strerror(fstatus));
	}
	else {
		fprintf(stdout,"Auks API request succeed\n");
	}

	auks_api_close(&engine);

exit:
	if ( conf_file_string != NULL )
		free(conf_file_string);

	return fstatus;
}
